<?xml version="1.0"?>

<launch>
    <param name="use_sim_time" value="true" type="bool"/>
    <arg name="map_file" default="$(find localizer)/config/map/map.yaml"/>
    <node pkg="map_server" name="map_server" type="map_server" args="$(arg map_file)" output="screen"/>
    <node pkg="rosbag" name="play" type="play" args="--clock $(env HOME)/bag_data/roomba1.bag $(env HOME)/bag_data/roomba2.bag $(env HOME)/bag_data/roomba3.bag $(env HOME)/bag_data/roomba4.bag $(env HOME)/bag_data/roomba5.bag $(env HOME)/bag_data/roomba6.bag $(env HOME)/bag_data/roomba2_positioning.bag $(env HOME)/bag_data/roomba3_positioning.bag $(env HOME)/bag_data/roomba4_positioning.bag $(env HOME)/bag_data/roomba5_positioning.bag $(env HOME)/bag_data/roomba6_positioning.bag"/>
    
    <group ns="roomba1">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="-1.0" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.5" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba0" type="string"/>
        <param name="reference_name2" value="/roomba2" type="string"/>
        <param name="parent" value="true" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba1/scan" />
        <remap from="/roomba/odometry" to="/roomba1/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba1/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba1/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba1/p_pose_array" />
        <remap from="/std_status" to="/roomba1/std_status" />
    </node>
    </group>
    
    <group ns="roomba2">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="-0.6" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.3" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba1" type="string"/>
        <param name="reference_name2" value="/roomba3" type="string"/>
        <param name="parent" value="false" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba2/scan" />
        <remap from="/target/position" to="/roomba2/target/position"/>
        <remap from="/roomba/odometry" to="/roomba2/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba2/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba2/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba2/p_pose_array" />
        <remap from="/std_status" to="/roomba2/std_status" />
    </node>
    </group>
    
    <group ns="roomba3">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="-0.2" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.3" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba2" type="string"/>
        <param name="reference_name2" value="/roomba4" type="string"/>
        <param name="parent" value="false" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba3/scan" />
        <remap from="/target/position" to="/roomba3/target/position"/>
        <remap from="/roomba/odometry" to="/roomba3/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba3/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba3/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba3/p_pose_array" />
        <remap from="/std_status" to="/roomba3/std_status" />
    </node>
    </group>
    
    <group ns="roomba4">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="0.2" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.3" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba3" type="string"/>
        <param name="reference_name2" value="/roomba5" type="string"/>
        <param name="parent" value="false" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba4/scan" />
        <remap from="/target/position" to="/roomba4/target/position"/>
        <remap from="/roomba/odometry" to="/roomba4/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba4/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba4/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba4/p_pose_array" />
        <remap from="/std_status" to="/roomba4/std_status" />
    </node>
    </group>
    
    <group ns="roomba5">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="0.6" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.8" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba4" type="string"/>
        <param name="reference_name2" value="/roomba6" type="string"/>
        <param name="parent" value="false" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba5/scan" />
        <remap from="/target/position" to="/roomba5/target/position"/>
        <remap from="/roomba/odometry" to="/roomba5/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba5/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba5/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba5/p_pose_array" />
        <remap from="/std_status" to="/roomba5/std_status" />
    </node>
    </group>

    <group ns="roomba6">
    <node pkg="localizer" type="Localizer" name="localizer" output="screen">
        <param name="hz" value="10" type="int"/>
        <param name="particle_number" value="500" type="int"/>
        <param name="init_x" value="1.0" type="double"/>
        <param name="init_y" value="7.0" type="double"/>
        <param name="init_yaw" value="3.14" type="double"/>
        <param name="init_x_sigma" value="0.3" type="double"/>
        <param name="init_y_sigma" value="0.3" type="double"/>
        <param name="init_yaw_sigma" value="0.3" type="double"/>
        <param name="move_noise_ratio" value="0.3" type="double"/>
        <param name="laser_noise_ratio" value="0.01" type="double"/>
        <param name="search_range" value="30.0" type="double"/>
        <param name="laser_step" value="10" type="int"/>
        <param name="alpha_slow_th" value="0.001" type="double"/>
        <param name="alpha_fast_th" value="0.1" type="double"/>
        <param name="reset_x_sigma" value="0.1" type="double"/>
        <param name="reset_y_sigma" value="0.1" type="double"/>
        <param name="reset_yaw_sigma" value="0.1" type="double"/>
        <param name="expansion_x_speed" value="0.05" type="double"/>
        <param name="expansion_y_speed" value="0.05" type="double"/>
        <param name="expansion_yaw_speed" value="0.02" type="double"/>
        <param name="estimated_pose_w_th" value="1.8" type="double"/>
        <param name="reset_limit" value="200" type="int"/>
        <param name="reference_name" value="/roomba5" type="string"/>
        <param name="reference_name2" value="/roomba7" type="string"/>
        <param name="parent" value="true" type="bool"/>
        <param name="observation_flag" value="true" type="bool"/>
        <remap from="/scan" to="/roomba6/scan" />
        <remap from="/target/position" to="/roomba6/target/position"/>
        <remap from="/roomba/odometry" to="/roomba6/roomba/odometry" />
        <remap from="/mcl_pose" to="/roomba6/mcl_pose" />
        <remap from="/move_base_simple/goal" to="/roomba6/move_base_simple/goal" />
        <remap from="/p_pose_array" to="/roomba6/p_pose_array" />
        <remap from="/std_status" to="/roomba6/std_status" />
    </node>
    </group>
     
    <node pkg="rviz" name="rviz" type="rviz" args="-d $(find localizer)/config/localizer.rviz" output="screen"/>
</launch>
